<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<!--[if IE]><meta http-equiv="X-UA-Compatible" content="IE=edge"><![endif]-->
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 1.5.6.1">
<title>Nextflow Hackaton 2017</title>
<link rel="stylesheet" href="./css/crg.css">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css">
<script type="text/javascript" src="https://code.jquery.com/jquery-1.12.4.js"></script>
</head>
<body class="article toc2 toc-left">
<div id="header">
<h1>Nextflow Hackaton 2017 <span class="image nflogo" style="float: right"><img src="./assets/nextflow_logo.png" alt="Logo" width="300" height="100"></span></h1>
<div id="toc" class="toc2">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel1">
<li><a href="#_data_description">1. Data Description</a></li>
<li><a href="#_worflow_description">2. Worflow Description</a>
<ul class="sectlevel2">
<li><a href="#_sofware_manuals">2.1. Sofware manuals</a></li>
<li><a href="#_pipeline_steps">2.2. Pipeline steps</a>
<ul class="sectlevel3">
<li><a href="#_preparing_data">2.2.1. Preparing data</a></li>
<li><a href="#_mapping_rna_seq_reads_to_the_reference">2.2.2. Mapping RNA-seq reads to the reference</a></li>
<li><a href="#_split_n_trim_and_reassign_mapping_qualities">2.2.3. Split&#8217;N&#8217;Trim and reassign mapping qualities</a></li>
<li><a href="#_base_recalibration">2.2.4. Base Recalibration</a></li>
<li><a href="#_variant_calling_and_variant_filtering">2.2.5. Variant Calling and Variant filtering</a></li>
<li><a href="#_variant_post_processing">2.2.6. Variant Post-processing</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_environment_setup">3. Environment Setup</a>
<ul class="sectlevel2">
<li><a href="#_launch_and_connect_to_the_ec2_virtual_machine">3.1. Launch and connect to the EC2 virtual machine</a></li>
<li><a href="#_installing_nextflow">3.2. Installing Nextflow</a></li>
<li><a href="#_getting_the_data">3.3. Getting the data</a></li>
<li><a href="#_pulling_the_docker_image">3.4. Pulling the Docker image</a></li>
</ul>
</li>
<li><a href="#_pipeline_implementaton">4. Pipeline Implementaton</a>
<ul class="sectlevel2">
<li><a href="#_data_preparation">4.1. Data preparation</a></li>
<li><a href="#_input_parameters">4.2. Input parameters</a>
<ul class="sectlevel3">
<li><a href="#_problem_1">4.2.1. Problem #1</a></li>
</ul>
</li>
<li><a href="#_process_1a">4.3. Process 1A</a>
<ul class="sectlevel3">
<li><a href="#_problem_2">4.3.1. Problem #2</a></li>
</ul>
</li>
<li><a href="#_process_1b">4.4. Process 1B</a>
<ul class="sectlevel3">
<li><a href="#_problem_3">4.4.1. Problem #3</a></li>
</ul>
</li>
<li><a href="#_process_1c">4.5. Process 1C</a>
<ul class="sectlevel3">
<li><a href="#_problem_4">4.5.1. Problem #4</a></li>
</ul>
</li>
<li><a href="#_process_1d">4.6. Process 1D</a>
<ul class="sectlevel3">
<li><a href="#_problem_5">4.6.1. Problem #5</a></li>
</ul>
</li>
<li><a href="#_process_2">4.7. Process 2</a>
<ul class="sectlevel3">
<li><a href="#_problem_6">4.7.1. Problem #6</a></li>
</ul>
</li>
<li><a href="#_process_3">4.8. Process 3</a>
<ul class="sectlevel3">
<li><a href="#_problem_7">4.8.1. Problem #7</a></li>
</ul>
</li>
<li><a href="#_process_4">4.9. Process 4</a>
<ul class="sectlevel3">
<li><a href="#_problem_8">4.9.1. Problem #8</a></li>
</ul>
</li>
<li><a href="#_process_5">4.10. Process 5</a>
<ul class="sectlevel3">
<li><a href="#_problem_9">4.10.1. Problem #9</a></li>
</ul>
</li>
<li><a href="#_processes_6a_and_6b">4.11. Processes 6A and 6B</a>
<ul class="sectlevel3">
<li><a href="#_problem_10">4.11.1. Problem #10</a></li>
<li><a href="#_problem_11">4.11.2. Problem #11</a></li>
</ul>
</li>
<li><a href="#_process_6c">4.12. Process 6C</a>
<ul class="sectlevel3">
<li><a href="#_problem_12">4.12.1. Problem #12</a></li>
</ul>
</li>
<li><a href="#_results_overview">4.13. Results overview</a></li>
<li><a href="#_bonus_step">4.14. Bonus step</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>This is the documentation for <a href="https://github.com/nextflow-io/hack17/issues/2" target="_blank" rel="noopener">Project 2</a> of the Nexflow Hackaton 2017. This project will cover the implementation of a Variant Calling analisys pipeline for RNAseq data based on GATK best practices and using Nextflow as the pipeline framework.</p>
</div>
<div class="paragraph">
<p>It is supposed to be a tutorial project for newbie Nextflow users who are interested in learning how to assemble a real world genomic pipeline with Nextflow, manage dependencies by using container technologies (Docker &amp; Singularity) and deploy the pipeline execution in a reproducible manner to the AWS cloud.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_data_description">1. Data Description</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The input data used to test the pipeline implementation is described below. For the purpose of this project, only a subset of the original data is used for most of the data types.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Genome assembly</dt>
<dd>
<p><code>genome.fa</code></p>
<div class="paragraph">
<p>The human genome assembly <span class="crg">hg19 (GRCh37)</span> from <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_000001405.1">GenBank</a>, chromosome 22 only.</p>
</div>
</dd>
<dt class="hdlist1">RNA-seq reads</dt>
<dd>
<p><code>ENCSR000COQ[12]_[12].fastq.gz</code></p>
<div class="paragraph">
<p>The RNA-seq data comes from the human <span class="crg">GM12878</span> cell line from whole cell, cytosol and nucleous extraction (see table below).</p>
</div>
<div class="paragraph">
<p>The libraries are <span class="crg">stranded PE76 Illumina GAIIx</span> RNA-Seq from <span class="crg">rRNA-depleted Poly-A+</span> long RNA (<code>&gt; 200</code> nucleotides in size).</p>
</div>
<div class="paragraph">
<p>Only reads mapped to the <a href="http://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;lastVirtModeType=default&amp;lastVirtModeExtraState=&amp;virtModeType=default&amp;virtMode=0&amp;nonVirtPosition=&amp;position=chr22%3A14700001-25900000&amp;hgsid=221945779_QucOFSFGagd1cn9uVki0TFjrxSBU" target="_blank" rel="noopener">22q11</a> locus of the human genome (<code>chr22:16000000-18000000</code>) are used.</p>
</div>
<table class="tableblock frame-all grid-all spread">
<colgroup>
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
</colgroup>
<thead>
<tr>
<th class="tableblock halign-left valign-top">ENCODE ID</th>
<th class="tableblock halign-left valign-top">Cellular fraction</th>
<th class="tableblock halign-left valign-top">replicate ID</th>
<th class="tableblock halign-left valign-top">file names</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock"><a href="https://www.encodeproject.org/experiments/ENCSR000COQ/">ENCSR000COQ</a></p></td>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock">Whole Cell</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">1</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000COQ1_1.fastq.gz
ENCSR000COQ1_2.fastq.gz</pre></div></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">2</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000COQ2_1.fastq.gz
ENCSR000COQ2_2.fastq.gz</pre></div></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock"><a href="https://www.encodeproject.org/experiments/ENCSR000CPO/">ENCSR000CPO</a></p></td>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock">Nuclear</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">1</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000CPO1_1.fastq.gz
ENCSR000CPO1_2.fastq.gz</pre></div></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">2</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000CPO2_1.fastq.gz
ENCSR000CPO2_2.fastq.gz</pre></div></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock"><a href="https://www.encodeproject.org/experiments/ENCSR000COR/">ENCSR000COR</a></p></td>
<td class="tableblock halign-left valign-top" rowspan="2"><p class="tableblock">Cytosolic</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">1</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000COR1_1.fastq.gz
ENCSR000COR1_1.fastq.gz</pre></div></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">2</p></td>
<td class="tableblock halign-left valign-top"><div class="literal"><pre>ENCSR000COR2_1.fastq.gz
ENCSR000COR2_1.fastq.gz</pre></div></td>
</tr>
</tbody>
</table>
</dd>
<dt class="hdlist1">"Known" variants</dt>
<dd>
<p><code>known_variants.vcf.gz</code></p>
<div class="paragraph">
<p>Known variants come from high confident variant calls for <span class="crg">GM12878</span> from the <a href="https://www.illumina.com/platinumgenomes.html">Illumina Platinum Genomes</a> project.
These variant calls were obtained by taking into account pedigree information and the concordance of calls across different methods.</p>
</div>
<div class="paragraph">
<p>We&#8217;re using the subset from chromosome 22 only.</p>
</div>
</dd>
<dt class="hdlist1">Blacklisted regions</dt>
<dd>
<p><code>blacklist.bed</code></p>
<div class="paragraph">
<p>Blacklisted regions are regions of the genomes with anomalous coverage. We use regions for the <span class="crg">hg19</span> assembly, taken from the <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/">ENCODE project portal</a>.
These regions were identified with DNAse and ChiP-seq samples over ~60 human tissues/cell types, and had a very high ratio of multi-mapping to unique-mapping reads and high variance in mappability.</p>
</div>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_worflow_description">2. Worflow Description</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The aim of the pipeline is to process raw RNA-seq data (in FASTQ format) and obtain the list of small variants, SNVs (SNPs and INDELs) for the downstream analysis. The pipeline is based on the <a href="https://software.broadinstitute.org/gatk/guide/article?id=3891">GATK best practices for variant calling with RNAseq data</a> and includes all major steps. In addition the pipeline includes SNVs postprocessing and quantification for allele specific expression.</p>
</div>
<div class="paragraph">
<p>Samples processing is done <strong>indipendently</strong> for <strong>each replicate</strong>. This includes mapping of the reads, splitting at the CIGAR, reassigning mapping qualities and recalibrating base qualities.</p>
</div>
<div class="paragraph">
<p>Variant calling is done <strong>simultaniously</strong> on bam files from <strong>all replicates</strong>. This allows to improve coverage of genomic regions and obtain more reliable results.</p>
</div>
<div class="sect2">
<h3 id="_sofware_manuals">2.1. Sofware manuals</h3>
<div class="paragraph">
<p>Documentation for all software used in the workflow can be found at the following links:</p>
</div>
<div class="ulist">
<ul>
<li>
<p><a href="http://www.htslib.org/doc/samtools.html">samtools</a></p>
</li>
<li>
<p><a href="https://broadinstitute.github.io/picard/command-line-overview.html#CreateSequenceDictionary">picard <code>CreateSequenceDictionary</code></a></p>
</li>
<li>
<p><a href="http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STAR.posix/doc/STARmanual.pdf">STAR</a></p>
</li>
<li>
<p><a href="https://vcftools.github.io/man_latest.html">vcftools</a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/index">GATK tools</a></p>
<div class="ulist">
<ul>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_rnaseq_SplitNCigarReads.php"><code>SplitNCigarReads</code></a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php"><code>BaseRecalibrator</code></a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_readutils_PrintReads.php"><code>PrintReads</code></a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php"><code>HaplotypeCaller</code></a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php"><code>VariantFiltration</code></a></p>
</li>
<li>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_rnaseq_ASEReadCounter.php"><code>ASEReadCounter</code></a></p>
</li>
</ul>
</div>
</li>
</ul>
</div>
</div>
<div class="sect2">
<h3 id="_pipeline_steps">2.2. Pipeline steps</h3>
<div class="paragraph">
<p>In order to get a general idea of the workflow, all the composing steps, together with the corresponding commands, are explained in the next sections.</p>
</div>
<div class="sect3">
<h4 id="_preparing_data">2.2.1. Preparing data</h4>
<div class="paragraph">
<p>This step prepares input files for the analysis. Genome indexes are created and variants overlapping blacklisted regions are filtered out.</p>
</div>
<div class="paragraph">
<p>Genome indices with <code>samtools</code> and <code>picard</code> are produced first. They will be needed for GATK commands such as <code>Split&#8217;N&#8217;Trim</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>samtools faidx genome.fa
java -jar picard.jar CreateSequenceDictionary R= genome.fa O= genome.dict</pre>
</div>
</div>
<div class="paragraph">
<p>Genome index for <code>STAR</code>, needed for RNA-seq reads mappings, is created next. Index files are written to the folder <code>genome_dir</code>
:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>STAR --runMode genomeGenerate \
     --genomeDir genome_dir \
     --genomeFastaFiles genome.fa \
     --runThreadN 4</pre>
</div>
</div>
<div class="paragraph">
<p>Variants overlapping blacklisted regions are then filtered in order to reduce false positive calls [optional]:</p>
</div>
<div class="literalblock">
<div class="content">
<pre>vcftools --gzvcf known_variants.vcf.gz -c \
         --exclude-bed blacklist.bed \
         --recode | bgzip -c \
         &gt; known_variants.filtered.recode.vcf.gz</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_mapping_rna_seq_reads_to_the_reference">2.2.2. Mapping RNA-seq reads to the reference</h4>
<div class="paragraph">
<p>To align RNA-seq reads to the genome we&#8217;re using STAR 2-pass approach. The first alignment creates a table with splice-junctions that is used to guide final alignments. The alignments at both steps are done with default parameters.</p>
</div>
<div class="paragraph">
<p>Additional fields with the read groups, libraries and sample information are added into the final bam file at the second mapping step. As a result we do not need to run Picard processing step from GATK best practices.</p>
</div>
<div class="paragraph">
<p>STAR 1-pass:</p>
</div>
<div class="listingblock">
<div class="content">
<pre> STAR --genomeDir genome_dir \
      --readFilesIn ENCSR000COQ1_1.fastq.gz ENCSR000COQ1_2.fastq.gz \
      --runThreadN 4 \
      --readFilesCommand zcat \
      --outFilterType BySJout \
      --alignSJoverhangMin 8 \
      --alignSJDBoverhangMin 1 \
      --outFilterMismatchNmax 999</pre>
</div>
</div>
<div class="paragraph">
<p>Create new genome index using splice-junction table:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>STAR --runMode genomeGenerate \
     --genomeDir genome_dir \
     --genomeFastaFiles genome.fa \
     --sjdbFileChrStartEnd SJ.out.tab \
     --sjdbOverhang 75 \
     --runThreadN 4</pre>
</div>
</div>
<div class="paragraph">
<p>STAR 2-pass, final alignments:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>  STAR --genomeDir genome_dir \
       --readFilesIn ENCSR000COQ1_1.fastq.gz ENCSR000COQ1_2.fastq.gz \
       --runThreadN 4 \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999 \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMattrRGline ID:ENCSR000COQ1 LB:library PL:illumina PU:machine SM:GM12878</pre>
</div>
</div>
<div class="paragraph">
<p>Index the resulting bam file:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>samtools index final_alignments.bam</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_split_n_trim_and_reassign_mapping_qualities">2.2.3. Split&#8217;N&#8217;Trim and reassign mapping qualities</h4>
<div class="paragraph">
<p>The RNA-seq reads overlapping exon-intron junctions can produce false positive variants due to inaccurate splicing. To solve this problem the GATK team recommend to hard-clip any sequence that overlap intronic regions and developed a speciall tool for this purpose: <code>SplitNCigarReads</code>. The tool identifies Ns in the CIGAR string of the alignment and split reads at this position so that few new reads are created.</p>
</div>
<div class="paragraph">
<p>At this step we also reassign mapping qualities to the alignments. This is important because STAR assign the value <code>255</code> (high quality) to “unknown” mappings that are meaningless to GATK and to variant calling in general.</p>
</div>
<div class="paragraph">
<p>This step is done with recommended parameters from the GATK best practices.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>java -jar GenomeAnalysisTK.jar -T SplitNCigarReads \
                  -R genome.fa -I final_alignments.bam \
                  -o split.bam \
                  -rf ReassignOneMappingQuality \
                  -RMQF 255 -RMQT 60 \
                  -U ALLOW_N_CIGAR_READS \
                  --fix_misencoded_quality_scores</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_base_recalibration">2.2.4. Base Recalibration</h4>
<div class="paragraph">
<p>The proposed worflow does not include an indel realigment step, which is an optional step in the GATK best practices. We excluded that since it is quite time-intensive and does not really improve variant calling.</p>
</div>
<div class="paragraph">
<p>We instead include a base recalibration step. This step allows to remove possible systematic errors introduced by the sequencing machine during the assignment of read qualities. To do this, the list of known variants is used as a training set to the machine learning algorithm that models possible errors. Base quality scores are then adjusted based on the obtained results.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>java -jar GenomeAnalysisTK.jar -T BaseRecalibrator \
                  --default_platform illumina \
                  -cov ReadGroupCovariate \
                  -cov QualityScoreCovariate \
                  -cov CycleCovariate \
                  -knownSites known_variants.filtered.recode.vcf.gz\
                  -cov ContextCovariate \
                  -R genome.fa -I split.bam \
                  --downsampling_type NONE \
                  -nct 4 \
                  -o final.rnaseq.grp

  java -jar GenomeAnalysisTK.jar -T PrintReads \
                  -R genome.fa -I split.bam \
                  -BQSR final.rnaseq.grp \
                  -nct 4 \
                  -o final.bam</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_variant_calling_and_variant_filtering">2.2.5. Variant Calling and Variant filtering</h4>
<div class="paragraph">
<p>The variant calling is done on the uniquely aligned reads only in order to reduce the number of false positive variants called:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>(samtools view -H final.bam; samtools view final.bam| grep -w 'NH:i:1') \
  | samtools view -Sb -  &gt; final.uniq.bam

samtools index final.uniq.bam</pre>
</div>
</div>
<div class="paragraph">
<p>For variant calling we&#8217;re using the GATK tool <code>HaplotypeCaller</code> with default parameters:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>ls final.uniq.bam  &gt; bam.list
java -jar GenomeAnalysisTK.jar -T HaplotypeCaller \
                  -R genome.fa -I bam.list \
                  -dontUseSoftClippedBases \
                  -stand_call_conf 20.0 \
                  -o output.gatk.vcf.gz</pre>
</div>
</div>
<div class="paragraph">
<p>Variant filtering is done as recommended in the GATK best practices:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>keep clusters of at least 3 SNPs that are within a window of 35 bases between them</p>
</li>
<li>
<p>estimate strand bias using Fisher&#8217;s Exact Test with values &gt; 30.0 (Phred-scaled p-value)</p>
</li>
<li>
<p>use variant call confidence score <code>QualByDepth</code> (QD) with values &lt; 2.0. The QD is the QUAL score normalized by allele depth (AD) for a variant.</p>
</li>
</ul>
</div>
<div class="listingblock">
<div class="content">
<pre> java -jar GenomeAnalysisTK.jar -T VariantFiltration \
                  -R genome.fa -V output.gatk.vcf.gz \
                  -window 35 -cluster 3 \
                  -filterName FS -filter "FS &gt; 30.0" \
                  -filterName QD -filter "QD &lt; 2.0" \
                  -o final.vcf</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_variant_post_processing">2.2.6. Variant Post-processing</h4>
<div class="paragraph">
<p>For downstream analysis we&#8217;re considering only sites that pass all filters and are covered with at least 8 reads:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>grep -v '#' final.vcf | awk '$7~/PASS/' \
| perl -ne 'chomp($_); ($dp)=$_=~/DP\\=(\\d+)\\;/; if($dp&gt;=8){print $_."\\n"};' &gt; result.DP8.vcf</pre>
</div>
</div>
<div class="paragraph">
<p>Filtered RNA-seq variants are compared with those obtained from DNA sequencing (from Illumina platinum genome project). Variants that are common to these two datasets are "known" SNVs. The ones present only in the RNA-seq cohort only are "novel".</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<div class="paragraph">
<p><strong>Known SNVs</strong> will be used for <strong>allele specific expression</strong> analysis.</p>
</div>
<div class="paragraph">
<p><strong>Novel variants</strong> will be used to detect <strong>RNA-editing events</strong>.</p>
</div>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>We compare two variants files to detect common and different sites:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>vcftools --vcf result.DP8.vcf --gzdiff known_SNVs.filtered.recode.vcf.gz --diff-site --out commonSNPs</pre>
</div>
</div>
<div class="paragraph">
<p>Here we select sites present in both files ("known" SNVs only):</p>
</div>
<div class="listingblock">
<div class="content">
<pre>awk 'BEGIN{OFS="\t"} $4~/B/{print $1,$2,$3}' commonSNPs.diff.sites_in_files  &gt; test.bed

vcftools --vcf final.vcf --bed test.bed --recode --keep-INFO-all --stdout &gt; known_snps.vcf</pre>
</div>
</div>
<div class="paragraph">
<p>Plot a histogram with allele frequency distribution for "known" SNVs:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>grep -v '#'  known_snps.vcf | awk -F '\\t' '{print $10}' \
               |awk -F ':' '{print $2}'|perl -ne 'chomp($_); \
               @v=split(/\\,/,$_); if($v[0]!=0 ||$v[1] !=0)\
               {print  $v[1]/($v[1]+$v[0])."\\n"; }' |awk '$1!=1' \
               &gt;AF.4R

gghist.R -i AF.4R -o AF.histogram.pdf</pre>
</div>
</div>
<div class="paragraph">
<p>Calculate read counts for each "known" SNVs per allele for allele specific expression analysis:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>java -jar GenomeAnalysisTK.jar -R genome.fa \
               -T ASEReadCounter \
               -o ASE.tsv \
               -I bam.list \
               -sites known_snps.vcf</pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_environment_setup">3. Environment Setup</h2>
<div class="sectionbody">
<div class="paragraph">
<p>We are going to use Amazon Elastic Compute Cloud (Amazon EC2) virtual machines to work on the practical part
of the workshop. Each attendee will run a <code>t2.xlarge</code> instance based on a prepared image with the following
tools already installed:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>Docker</p>
</li>
<li>
<p>Git</p>
</li>
<li>
<p>Nano</p>
</li>
<li>
<p>Java 8</p>
</li>
</ul>
</div>
<div class="sect2">
<h3 id="_launch_and_connect_to_the_ec2_virtual_machine">3.1. Launch and connect to the EC2 virtual machine</h3>
<div class="paragraph">
<p>Connect to the AWS cloud login instance by using the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">ssh login@34.252.30.18</code></pre>
</div>
</div>
<div class="paragraph">
<p>Then launch a new AWS instance running the command shown below and follow
the instructions on screen:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">curl -L https://goo.gl/1C3abb | bash</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>~~ N E X T F L O W  '1 7  -  W O R K S H O P ~~

Launching EC2 virtual machine
- type  : t2.xlarge
- ami   : ami-61d0ec07
- disk  : 8 GB
- subnet: subnet-05222a43

* Please confirm you want to launch an VM with these settings [y/n]</pre>
</div>
</div>
<div class="paragraph">
<p>Type <kbd>Y</kbd> <kbd>Enter</kbd> and wait until the instance is ready:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>* Instance launched &gt;&gt; i-07a9f2ba3795c4f8f &lt;&lt;
* Waiting for ready status ..  [/]</pre>
</div>
</div>
<div class="paragraph">
<p>Once the previous command is completed and the virtual machine is up and
running you will get a message similar to the one below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>* The instance is ready -- Login with the following command:

  ssh hack17@ec2-xx-xxx-xxx-xxx.eu-west-1.compute.amazonaws.com</pre>
</div>
</div>
<div class="paragraph">
<p>Open a new shell terminal and connect to the AWS instance using the
command printed.</p>
</div>
</div>
<div class="sect2">
<h3 id="_installing_nextflow">3.2. Installing Nextflow</h3>
<div class="paragraph">
<p>Nextflow can be installed with the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">curl -fsSL get.nextflow.io | bash</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>...
CAPSULE: Downloading dependency com.fasterxml.jackson.core:jackson-databind:jar:2.6.6
CAPSULE: Downloading dependency ch.qos.logback:logback-classic:jar:1.1.10
CAPSULE: Downloading dependency com.amazonaws:jmespath-java:jar:1.0
CAPSULE: Downloading dependency com.fasterxml.jackson.core:jackson-annotations:jar:2.6.0

      N E X T F L O W
      version 0.25.6 build 4526
      last modified 13-08-2017 14:35 UTC
      cite doi:10.1038/nbt.3820
      http://nextflow.io


Nextflow installation completed. Please note:
- the executable file `nextflow` has been created in the folder: /home/hack17
- you may complete the installation by moving it to a directory in your $PATH</pre>
</div>
</div>
<div class="paragraph">
<p>Complete the installation copying the <code>nextflow</code> launcher file in a folder in the <code>PATH</code> variable e.g.:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">mv nextflow $HOME/bin/</code></pre>
</div>
</div>
<div class="paragraph">
<p>To check the installed version use the command below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow info</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>  Version: 0.25.6 build 4526
  Modified: 13-08-2017 14:35 UTC
  System: Linux 4.4.0-93-generic
  Runtime: Groovy 2.4.11 on OpenJDK 64-Bit Server VM 1.8.0_131-8u131-b11-2ubuntu1.16.04.3-b11
  Encoding: UTF-8 (UTF-8)</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_getting_the_data">3.3. Getting the data</h3>
<div class="paragraph">
<p>All the files needed for the workshop are stored in a git repository on GitHub. In order to get the data run the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">git clone https://github.com/CRG-CNAG/ngs2017-nf.git</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>Cloning into 'ngs2017-nf'...
remote: Counting objects: 71, done.
remote: Total 71 (delta 0), reused 0 (delta 0), pack-reused 71
Unpacking objects: 100% (71/71), done.
Checking connectivity... done.</pre>
</div>
</div>
<div class="paragraph">
<p>Check the content of the freshly cloned repo:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">tree ngs2017-nf</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>ngs2017-nf
├── bin
│   └── gghist.R
├── data
│   ├── blacklist.bed
│   ├── genome.fa
│   ├── known_variants.vcf.gz
│   └── reads
│       ├── ENCSR000COQ1_1.fastq.gz
│       ├── ENCSR000COQ1_2.fastq.gz
│       ├── ENCSR000COQ2_1.fastq.gz
│       ├── ENCSR000COQ2_2.fastq.gz
│       ├── ENCSR000COR1_1.fastq.gz
│       ├── ENCSR000COR1_2.fastq.gz
│       ├── ENCSR000COR2_1.fastq.gz
│       ├── ENCSR000COR2_2.fastq.gz
│       ├── ENCSR000CPO1_1.fastq.gz
│       ├── ENCSR000CPO1_2.fastq.gz
│       ├── ENCSR000CPO2_1.fastq.gz
│       └── ENCSR000CPO2_2.fastq.gz
├── launch-ec2.sh
├── nextflow.config
└── README.md

3 directories, 19 files</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_pulling_the_docker_image">3.4. Pulling the Docker image</h3>
<div class="paragraph">
<p>Nextflow can pull Docker images at runtime, but let&#8217;s just download it manually to see how Docker works:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">docker pull cbcrg/callings-nf@sha256:b65a7d721b9dd2da07d6bdd7f868b04039860f14fa514add975c59e68614c310</code></pre>
</div>
</div>
<div class="paragraph">
<p>You should see the progress of the download:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>sha256:b65a7d721b9dd2da07d6bdd7f868b04039860f14fa514add975c59e68614c310: Pulling from cbcrg/callings-nf
915665fee719: Downloading [=============================================&gt;     ] 47.08 MB/51.36 MB
f332de2321e6: Downloading [===========&gt;                                       ] 41.96 MB/187.8 MB
1577a6dd9e43: Downloading [===============================&gt;                   ] 46.72 MB/73.45 MB
7059d9bb5245: Waiting
71863f70269f: Waiting
ce2a2879246d: Waiting
e38ba5d5f9fb: Waiting
90158da87bb2: Waiting</pre>
</div>
</div>
<div class="paragraph">
<p>and the following message when the pull is completed:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>Digest: sha256:b65a7d721b9dd2da07d6bdd7f868b04039860f14fa514add975c59e68614c310
Status: Downloaded newer image for cbcrg/callings-nf@sha256:b65a7d721b9dd2da07d6bdd7f868b04039860f14fa514add975c59e68614c310</pre>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_pipeline_implementaton">4. Pipeline Implementaton</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_data_preparation">4.1. Data preparation</h3>
<div class="paragraph">
<p>A first step in any pipeline is to prepare the input data. You will find
all the data required to run the pipeline in the folder <code>data</code>
within the <code>ngs2017-nf</code> repository directory.</p>
</div>
<div class="paragraph">
<p>There are four data inputs that we will use in this tutorial:</p>
</div>
<div class="olist arabic">
<ol class="arabic">
<li>
<p><strong>Genome File</strong> (<code>data/genome.fa</code>)</p>
<div class="ulist">
<ul>
<li>
<p>Human chromosome 22 in FASTA file format</p>
</li>
</ul>
</div>
</li>
<li>
<p><strong>Read Files</strong> (<code>data/reads/</code>)</p>
<div class="ulist">
<ul>
<li>
<p>Sample ENCSR000COQ1: 76bp paired-end reads (<code>ENCSR000COQ1_1.fq.gz</code> and <code>ENCSR000COQ1_2.fq.gz</code>).</p>
</li>
</ul>
</div>
</li>
<li>
<p><strong>Variants File</strong> (<code>data/known_variants.vcf.gz</code>)</p>
<div class="ulist">
<ul>
<li>
<p>Known variants, gzipped as a Variant Calling File (VCF) format.</p>
</li>
</ul>
</div>
</li>
<li>
<p><strong>Blacklist File</strong> (<code>data/blacklist.bed</code>)</p>
<div class="ulist">
<ul>
<li>
<p>Genomic locations which are known to produce artifacts and spurious variants in Browser Extensible Data (BED) format.</p>
</li>
</ul>
</div>
</li>
</ol>
</div>
<hr>
</div>
<div class="sect2">
<h3 id="_input_parameters">4.2. Input parameters</h3>
<div class="paragraph">
<p>We can begin writing the pipeline by creating and editing a text file called <code>main.nf</code>
from the <code>ngs2017-nf</code> repository directory with your favourite text editor. In this example we are using <code>nano</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">cd ngs2017-nf
nano main.nf</code></pre>
</div>
</div>
<div class="paragraph">
<p>Edit this file to specify the input files as script parameters. Using this notation
allows you to override them by specifying different values when launching the
pipeline execution.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Define the default parameters <i class="conum" data-value="1"></i><b>(1)</b>
 */

params.genome     = "data/genome.fa"
params.variants   = "data/known_variants.vcf.gz"
params.blacklist  = "data/blacklist.bed"
params.reads      = "data/reads/ENCSR000COQ1_{1,2}.fastq.gz" <i class="conum" data-value="2"></i><b>(2)</b>
params.results    = "results" <i class="conum" data-value="3"></i><b>(3)</b>
params.gatk       = "/opt/broad/GenomeAnalysisTK.jar" <i class="conum" data-value="4"></i><b>(4)</b></pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
You can copy the above text by using the <span class="keyseq"><kbd>Cmd</kbd>+<kbd>C</kbd></span> keys, then move in the terminal window,
open <code>nano</code> and paste the above text by using the <span class="keyseq"><kbd>Cmd</kbd>+<kbd>V</kbd></span> keys shortcut.
</td>
</tr>
</table>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The <code>/*</code>, <code>*</code> and <code>*/</code> specify comment lines which are ignored by Nextflow.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The <code>reads</code> parameter uses a glob pattern to specify the forward (<code>ENCSR000COQ1_1.fq.gz</code>) and reverse (<code>ENCSR000COQ1_2.fq.gz</code>) reads are pairs of the same sample.</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>The <code>results</code> parameter is used to specify a directory called <code>results</code>.</td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td>The <code>gatk</code> parameter specifies the location of the GATK jar file.</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Once you have the default parameters in the <code>main.nf</code> file, you can save and run the main script for the first time.</p>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
With <code>nano</code> you can save and close the file with <span class="keyseq"><kbd>Ctrl</kbd>+<kbd>O</kbd></span>, then <kbd>Enter</kbd>, followed by <span class="keyseq"><kbd>Ctrl</kbd>+<kbd>X</kbd></span>.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>To run the main script use the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow run main.nf</code></pre>
</div>
</div>
<div class="paragraph">
<p>You should see the script execute, print Nextflow version and pipeline revision and then exit.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>N E X T F L O W  ~  version 0.24.1
Launching `main.nf` [nauseous_wright] - revision: 83a0a5415a</pre>
</div>
</div>
<hr>
<div class="sect3">
<h4 id="_problem_1">4.2.1. Problem #1</h4>
<div class="paragraph">
<p>Great, now we need to transform the parameters into proper file handlers and <a href="https://www.nextflow.io/docs/latest/channel.html">channel</a> variables.
To do that open the <code>main.nf</code> file and copy the lines below at the end of the file.</p>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
In <code>nano</code> you can move to the end of the file using <span class="keyseq"><kbd>Ctrl</kbd>+<kbd>W</kbd></span> and then <span class="keyseq"><kbd>Ctrl</kbd>+<kbd>V</kbd></span>.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>This time you must fill the <code>BLANK</code> spaces with the correct function and parameter.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 *  Parse the input parameters
 */

genome_file     = file(params.genome)
variants_file   = BLANK
blacklist_file  = BLANK
reads_ch        = BLANK
GATK            = params.gatk</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
The first three should specify file objects containing a single <code>file</code> as shown in
<a href="https://www.nextflow.io/docs/latest/basic.html#processes-and-channels">this</a> basic example.
For the reads channel, use the <a href="https://www.nextflow.io/docs/latest/channel.html#fromfilepairs">fromFilePairs</a>
channel factory method. The final one, <code>GATK</code> is simply creating a Nextflow variable
specifying the path of the GATK application file and does not need any special function.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Once you think you have data organised, you can again run the pipeline.
However this time, we can use the the <code>-resume</code> flag.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow run main.nf -resume</code></pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
See <a href="https://www.nextflow.io/docs/latest/getstarted.html?highlight=resume#modify-and-resume">here</a> for more
details about using the <code>resume</code> option.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/anxious_advice.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_process_1a">4.3. Process 1A</h3>
<h3 id="_create_a_fasta_genome_index" class="discrete">Create a FASTA genome index</h3>
<div class="paragraph">
<p>Now we have our inputs set up we can move onto the processes. In our first process we will
create a genome index using <a href="http://www.htslib.org/">samtools</a>.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>1A_prepare_genome_samtools</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>create a genome index for the genome fasta with samtools</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>the samtools genome index file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_2">4.3.1. Problem #2</h4>
<div class="paragraph">
<p>Copy the code below and paste it at the end of <code>main.nf</code>.</p>
</div>
<div class="paragraph">
<p>Your aim is to replace <code>BLANK</code> placeholder with the  the correct
variable name of the genome file that you have defined in previous problem.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Process 1A: Create a FASTA genome index with samtools
 */

process '1A_prepare_genome_samtools' { <i class="conum" data-value="1"></i><b>(1)</b>

  input:
      file genome from BLANK <i class="conum" data-value="2"></i><b>(2)</b>

  output:
      file "${genome}.fai" into genome_index_ch <i class="conum" data-value="3"></i><b>(3)</b>

  script:
  """
  samtools faidx ${genome} <i class="conum" data-value="4"></i><b>(4)</b>
  """
}</pre>
</div>
</div>
<div class="paragraph">
<p>In plain english, the process could be written as:</p>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>A <strong>process</strong> called 1A_prepare_genome_samtools</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>takes as <strong>input</strong> the genome file from <code>BLANK</code></td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>and creates as <strong>output</strong> a genome index file which goes into channel <code>genome_index_ch</code></td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td><strong>script</strong>: using samtools create the genome index from the genome file</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Now when we run the pipeline, we see that the process 1A is submitted:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow run main.nf -resume</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre>N E X T F L O W  ~  version 0.24.1
Launching `main.nf` [adoring_wilson] - revision: 89dbc97b8e
[warm up] executor &gt; local
[17/b0eae4] Submitted process &gt; 1A_prepare_genome_samtools</pre>
</div>
</div>
<div class="paragraph">
<p><a href="solutions/busy_building.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_process_1b">4.4. Process 1B</h3>
<h3 id="_create_a_fasta_genome_sequence_dictionary_with_picard_for_gatk" class="discrete">Create a FASTA genome sequence dictionary with Picard for GATK</h3>
<div class="paragraph">
<p>Our first process created the genome index for GATK using samtools. For the next process we must do something very similar, this time creating a genome sequence dictionary using <a href="https://broadinstitute.github.io/picard/">Picard</a>.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>1B_prepare_genome_picard</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>create a genome dictionary for the genome fasta with Picard tools</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>the genome dictionary file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_3">4.4.1. Problem #3</h4>
<div class="paragraph">
<p>Fill in the <code>BLANK</code> words for both the input and output sections.</p>
</div>
<div class="paragraph">
<p>Copy the code below and paste it at the end of <code>main.nf</code>.</p>
</div>
<div class="paragraph">
<p>Your aim is to insert the correct input name from into
the input step (written as <code>BLANK</code>) of the process and run the pipeline.</p>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
You can choose any channel output name that makes sense to you.
</td>
</tr>
</table>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK
 */

process '1B_prepare_genome_picard' {

  input:
      file genome BLANK BLANK

  output:
      file "${genome.baseName}.dict" BLANK BLANK

  script:
  """
  PICARD=`which picard.jar`
  java -jar \$PICARD CreateSequenceDictionary R= $genome O= ${genome.baseName}.dict
  """
}</pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<code>.baseName</code> returns the filename without the file suffix. If <code>"${genome}"</code> is <code>human.fa</code>, then <code>"${genome.baseName}.dict"</code> would be <code>human.dict</code>.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/cooing_clock.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_process_1c">4.5. Process 1C</h3>
<h3 id="_create_star_genome_index_file" class="discrete">Create STAR genome index file</h3>
<div class="paragraph">
<p>Next we must create a genome index for the <a href="https://github.com/alexdobin/STAR">STAR</a> mapping software.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>1C_prepare_star_genome_index</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>create a STAR genome index for the genome fasta</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a directory containing the STAR genome index</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_4">4.5.1. Problem #4</h4>
<div class="paragraph">
<p>This is a similar exercise as problem 3, except this time both <code>input</code> and <code>output</code> lines have been left <code>BLANK</code> and must be completed.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Process 1C: Create the genome index file for STAR
 */

process '1C_prepare_star_genome_index' {

  input:
      BLANK_LINE

  output:
      BLANK_LINE

  script:
  """
  mkdir genome_dir

  STAR --runMode genomeGenerate \
       --genomeDir genome_dir \
       --genomeFastaFiles ${genome} \
       --runThreadN ${task.cpus}
  """
}</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
The output of the STAR genomeGenerate command is specified here as <code>genome_dir</code>.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/discreet_direction.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_process_1d">4.6. Process 1D</h3>
<h3 id="_filtered_and_recoded_set_of_variants" class="discrete">Filtered and recoded set of variants</h3>
<div class="paragraph">
<p>Next on to something a little more tricky. The next process takes two inputs: the variants
file and the blacklist file.</p>
</div>
<div class="paragraph">
<p>It should output a channel named <code>prepared_vcf_ch</code> which emitting a tuple of two files.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
In Nextflow, tuples can be defined in the input or output using the <a href="https://www.nextflow.io/docs/latest/process.html?highlight=set#output-set-of-values"><code>set</code></a> qualifier.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>1D_prepare_vcf_file</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>create a filtered and recoded set of variants</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the variants file<br>
the blacklisted regions file</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the filtered/recoded VCF file and the tab index (TBI) file.</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_5">4.6.1. Problem #5</h4>
<div class="paragraph">
<p>You must fill in the two <code>BLANK_LINES</code> in the input and the two <code>BLANK</code> output files.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Process 1D: Create a file containing the filtered and recoded set of variants
 */

process '1D_prepare_vcf_file' {

  input:
      BLANK_LINE
      BLANK_LINE

  output:
      set BLANK, BLANK into prepared_vcf_ch

  script:
  """
  vcftools --gzvcf $variantsFile -c \<i class="conum" data-value="1"></i><b>(1)</b>
           --exclude-bed ${blacklisted} \<i class="conum" data-value="2"></i><b>(2)</b>
           --recode | bgzip -c \
           &gt; ${variantsFile.baseName}.filtered.recode.vcf.gz <i class="conum" data-value="3"></i><b>(3)</b>

  tabix ${variantsFile.baseName}.filtered.recode.vcf.gz <i class="conum" data-value="4"></i><b>(4)</b>
  """
}</pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The input variable for the variants file</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The input variable for the blacklist file</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>The first of the two output files</td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td>Generates the second output file named <code>"${variantsFile.baseName}.filtered.recode.vcf.gz.tbi"</code></td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Try run the pipeline from the project directory with:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow run main.nf -resume</code></pre>
</div>
</div>
<div class="paragraph">
<p><a href="solutions/expensive_ear.html">Solution</a></p>
</div>
<div class="paragraph">
<p>Congratulations! Part 1 is now complete.</p>
</div>
<hr>
<div class="paragraph">
<p>We have all the data prepared and into channels ready for the more serious steps</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_process_2">4.7. Process 2</h3>
<h3 id="_star_mapping" class="discrete">STAR Mapping</h3>
<div class="paragraph">
<p>In this process, for each sample, we align the reads to our genome using the STAR index we created previously.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>2_rnaseq_mapping_star</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>mapping of the RNA-Seq reads using STAR</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file<br>
the STAR genome index<br>
a set containing the replicate id and paired read files</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containg replicate id, aligned bam file &amp; aligned bam file index</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_6">4.7.1. Problem #6</h4>
<div class="paragraph">
<p>Copy the code below and paste it at the end of <code>main.nf</code>.</p>
</div>
<div class="paragraph">
<p>You must fill in the three <code>BLANK_LINE</code> lines in the input and the one <code>BLANK_LINE</code> line in the output.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>/*
 * Process 2: Align RNA-Seq reads to the genome with STAR
 */

process '2_rnaseq_mapping_star' {

  input:
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE

  output:
      BLANK_LINE

  script:
  """
  # ngs-nf-dev Align reads to genome
  STAR --genomeDir $genomeDir \
       --readFilesIn $reads \
       --runThreadN ${task.cpus} \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999

  # 2nd pass (improve alignmets using table of splice junctions and create a new index)
  mkdir genomeDir
  STAR --runMode genomeGenerate \
       --genomeDir genomeDir \
       --genomeFastaFiles $genome \
       --sjdbFileChrStartEnd SJ.out.tab \
       --sjdbOverhang 75 \
       --runThreadN ${task.cpus}

  # Final read alignments
  STAR --genomeDir genomeDir \
       --readFilesIn $reads \
       --runThreadN ${task.cpus} \
       --readFilesCommand zcat \
       --outFilterType BySJout \
       --alignSJoverhangMin 8 \
       --alignSJDBoverhangMin 1 \
       --outFilterMismatchNmax 999 \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMattrRGline ID:$replicateId LB:library PL:illumina PU:machine SM:GM12878

  # Index the BAM file
  samtools index Aligned.sortedByCoord.out.bam
  """
}</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
The final command produces an bam index which is the full filename with an additional <code>.bai</code> suffix.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/fat_floor.html">Solution</a></p>
</div>
<hr>
<div class="paragraph">
<p>The next step is a filtering step using GATK. For each sample, we split all the reads that contain
N characters in their <a href="http://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F">CIGAR</a> string.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_process_3">4.8. Process 3</h3>
<h3 id="_gatk_split_on_n" class="discrete">GATK Split on N</h3>
<div class="paragraph">
<p>The process creates k+1 new reads (where k is the number of N cigar elements)
that correspond to the segments of the original read beside/between
the splicing events represented by the Ns in the original CIGAR.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>3_rnaseq_gatk_splitNcigar</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>split reads on Ns in CIGAR string using GATK</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file<br>
the genome index made with samtools<br>
the genome dictionary made with picard<br>
a set containg replicate id, aligned bam file and aligned bam file index from the STAR mapping</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the sample id, the split bam file and the split bam index file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_7">4.8.1. Problem #7</h4>
<div class="paragraph">
<p>Copy the code below and paste it at the end of <code>main.nf</code>.</p>
</div>
<div class="paragraph">
<p>You must fill in the four <code>BLANK_LINE</code> lines in the input and the one <code>BLANK_LINE</code> line in the output.</p>
</div>
<div class="admonitionblock caution">
<table>
<tr>
<td class="icon">
<i class="fa icon-caution" title="Caution"></i>
</td>
<td class="content">
There is an optional <a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line added
to the start of this process. The <a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line
allows you to assign a name to a specific task (single execution of a process).
This is particularly useful when there are many samples/replicates which pass through the same process.
</td>
</tr>
</table>
</div>
<div class="listingblock">
<div class="content">
<pre>process '3_rnaseq_gatk_splitNcigar' {
  tag OPTIONAL_BLANK

  input:
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE

  output:
      BLANK_LINE

  script:
  """
  # SplitNCigarReads and reassign mapping qualities
  java -jar $GATK -T SplitNCigarReads \
                  -R $genome -I $bam \
                  -o split.bam \
                  -rf ReassignOneMappingQuality \
                  -RMQF 255 -RMQT 60 \
                  -U ALLOW_N_CIGAR_READS \
                  --fix_misencoded_quality_scores
  """
}</pre>
</div>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
The GATK command above automatically creates a bam index (.bai) of the split.bam output file
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/gentle_garden.html">Solution</a></p>
</div>
<hr>
<div class="paragraph">
<p>Next we perform a Base Quality Score Recalibration step using GATK.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_process_4">4.9. Process 4</h3>
<h3 id="_gatk_recalibrate" class="discrete">GATK Recalibrate</h3>
<div class="paragraph">
<p>This step uses GATK to detect systematic errors in the base quality scores, select unique alignments and then index the resulting bam file with samtools. You can find details of the specific GATK BaseRecalibrator parameters <a href="https://software.broadinstitute.org/gatk/gatkdocs/3.6-0/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php">here</a>.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>4_rnaseq_gatk_recalibrate</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>recalibrate reads from each replicate using GATK</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file<br>
the genome index made with samtools<br>
the genome dictionary made with picard<br>
a set containg replicate id, aligned bam file and aligned bam file index from process 3<br>
a set containing the filtered/recoded VCF file and the tab index (TBI) file from process 1D<br></p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the sample id, the unique bam file and the unique bam index file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_8">4.9.1. Problem #8</h4>
<div class="paragraph">
<p>Copy the code below and paste it at the end of <code>main.nf</code>.</p>
</div>
<div class="paragraph">
<p>You must fill in the five <code>BLANK_LINE</code> lines in the input and the one <code>BLANK_LINE</code> line in the output.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>process '4_rnaseq_gatk_recalibrate' {
  tag "$replicateId"

  input:
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE

  output:
      BLANK into (final_output_ch, bam_for_ASE_ch) <i class="conum" data-value="1"></i><b>(1)</b>

  script:
    sampleId = replicateId.replaceAll(/[12]$/,'')
    """
    # Indel Realignment and Base Recalibration
    java -jar $GATK -T BaseRecalibrator \
                  --default_platform illumina \
                  -cov ReadGroupCovariate \
                  -cov QualityScoreCovariate \
                  -cov CycleCovariate \
                  -knownSites ${variants_file} \
                  -cov ContextCovariate \
                  -R ${genome} -I ${bam} \
                  --downsampling_type NONE \
                  -nct ${task.cpus} \
                  -o final.rnaseq.grp

     java -jar $GATK -T PrintReads \
                  -R ${genome} -I ${bam} \
                  -BQSR final.rnaseq.grp \
                  -nct ${task.cpus} \
                  -o final.bam

    # Select only unique alignments, no multimaps
    (samtools view -H final.bam; samtools view final.bam| grep -w 'NH:i:1') \
    |samtools view -Sb -  &gt; ${replicateId}.final.uniq.bam <i class="conum" data-value="2"></i><b>(2)</b>

    # Index BAM files
    samtools index ${replicateId}.final.uniq.bam <i class="conum" data-value="3"></i><b>(3)</b>
    """
}</pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>The files resulting from this process will be used in two downstream processes. If a process is executed more than once, and the downstream channel is used by more than one process, we must duplicate the channel. We can do this using the <code>into</code> operator with parenthesis in the output section. See <a href="https://www.nextflow.io/docs/latest/operator.html#into">here</a> for more information on using <code>into</code>.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>The unique bam file</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>The index of the unique bam file (bam file name + <code>.bai</code>)</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/hulking_hospital.html">Solution</a></p>
</div>
<hr>
<div class="paragraph">
<p>Now we are ready to perform the variant calling with GATK.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_process_5">4.10. Process 5</h3>
<h3 id="_gatk_variant_calling" class="discrete">GATK Variant Calling</h3>
<div class="paragraph">
<p>This steps call variants with GATK HaplotypeCaller. You can find details of the specific
GATK HaplotypeCaller parameters <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">here</a>.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>5_rnaseq_call_variants</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>variant calling of each sample using GATK</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>the genome fasta file<br>
the genome index made with samtools<br>
the genome dictionary made with picard<br>
a set containg replicate id, aligned bam file and aligned bam file index from process 4</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the sample id the resulting variant calling file (vcf)</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_9">4.10.1. Problem #9</h4>
<div class="paragraph">
<p>In this problem we will introduce the use of a channel operator in the input section.
The <a href="https://www.nextflow.io/docs/latest/operator.html#grouptuple">groupTuple</a> operator
groups together the tuples emitted by a channel which share a common key.</p>
</div>
<div class="admonitionblock caution">
<table>
<tr>
<td class="icon">
<i class="fa icon-caution" title="Caution"></i>
</td>
<td class="content">
Note that in process 4, we used the sampleID (not replicateID) as the first element
of the set in the output. Now we combine the replicates by grouping them on the sample ID.
It follows from this that process 4 is run one time per replicate and process 5 is run one time per sample.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Fill in the <code>BLANKS</code> as before.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>process '5_rnaseq_call_variants' {
  tag BLANK

  input:
      BLANK_LINE
      BLANK_LINE
      BLANK_LINE
      BLANK from BLANK.groupTuple()

  output:
      BLANK_LINE

  script:
  """
  echo "${bam.join('\n')}" &gt; bam.list

  # Variant calling
  java -jar $GATK -T HaplotypeCaller \
                  -R $genome -I bam.list \
                  -dontUseSoftClippedBases \
                  -stand_call_conf 20.0 \
                  -o output.gatk.vcf.gz

  # Variant filtering
  java -jar $GATK -T VariantFiltration \
                  -R $genome -V output.gatk.vcf.gz \
                  -window 35 -cluster 3 \
                  -filterName FS -filter "FS &gt; 30.0" \
                  -filterName QD -filter "QD &lt; 2.0" \
                  -o final.vcf
  """
}</pre>
</div>
</div>
<div class="paragraph">
<p><a href="solutions/imported_iron.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_processes_6a_and_6b">4.11. Processes 6A and 6B</h3>
<h3 id="_ase_rna_editing" class="discrete">ASE &amp; RNA Editing</h3>
<div class="paragraph">
<p>In the final steps we will create processes for Allele-Specific Expression and RNA Editing Analysis.</p>
</div>
<div class="paragraph">
<p>We must process the VCF result to prepare variants file for allele specific expression (ASE) analysis. We will implement both processes togther.</p>
</div>
<div class="paragraph">
<p>You should implement two processes having the following structure:</p>
</div>
<div class="dlist">
<div class="title">1st process</div>
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>6A_post_process_vcf</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>post-process the variant calling file (vcf) of each sample</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>set containing the sample ID and vcf file<br>
a set containing the filtered/recoded VCF file and the tab index (TBI) file from process 1D<br></p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the sample id, the variant calling file (vcf) and a file containing common SNPs</p>
</dd>
</dl>
</div>
<div class="dlist">
<div class="title">2nd process</div>
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>6B_prepare_vcf_for_ase</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>prepare the VCF for allele specific expression (ASE) and generate a figure in R.</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>a set containing the sample id, the variant calling file (vcf) and a file containing common SNPs</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>a set containing the sample ID and known SNPs in the sample for ASE<br>
a figure of the SNPs generated in R as a PDF file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_10">4.11.1. Problem #10</h4>
<div class="paragraph">
<p>Here we introduce the <code>publishDir</code> directive. This allows us to specifiy a location for the outputs of the process. See <a href="https://www.nextflow.io/docs/latest/process.html#publishdir">here</a> for more details.</p>
</div>
<div class="paragraph">
<p>You must have the output of process 6A become the input of process 6B.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>process '6A_post_process_vcf' {
  tag BLANK
  publishDir "$params.results/$sampleId" <i class="conum" data-value="1"></i><b>(1)</b>

  input:
      BLANK_LINE
      BLANK_LINE

  output:
      BLANK_LINE

  script:
  '''
  grep -v '#' final.vcf | awk '$7~/PASS/' |perl -ne 'chomp($_); ($dp)=$_=~/DP\\=(\\d+)\\;/; if($dp&gt;=8){print $_."\\n"};' &gt; result.DP8.vcf

  vcftools --vcf result.DP8.vcf --gzdiff filtered.recode.vcf.gz  --diff-site --out commonSNPs
  '''
}


process '6B_prepare_vcf_for_ase' {
  tag BLANK
  publishDir BLANK

  input:
      BLANK_LINE
  output:
      BLANK_LINE
      BLANK_LINE

  script:
  '''
  awk 'BEGIN{OFS="\t"} $4~/B/{print $1,$2,$3}' commonSNPs.diff.sites_in_files  &gt; test.bed

  vcftools --vcf final.vcf --bed test.bed --recode --keep-INFO-all --stdout &gt; known_snps.vcf

  grep -v '#'  known_snps.vcf | awk -F '\\t' '{print $10}' \
               |awk -F ':' '{print $2}'|perl -ne 'chomp($_); \
               @v=split(/\\,/,$_); if($v[0]!=0 ||$v[1] !=0)\
               {print  $v[1]/($v[1]+$v[0])."\\n"; }' |awk '$1!=1' \
               &gt;AF.4R

  gghist.R -i AF.4R -o AF.histogram.pdf
  '''
}</pre>
</div>
</div>
<div class="paragraph">
<p><a href="solutions/jumping_jack.html">Solution</a></p>
</div>
<hr>
<div class="paragraph">
<p>The final step is the GATK ASEReadCounter.</p>
</div>
</div>
<div class="sect3">
<h4 id="_problem_11">4.11.2. Problem #11</h4>
<div class="paragraph">
<p>We have seen the basics of using processes in Nextflow. Yet one of the standout
features of Nextflow is the operations that can be performed on
channels outside of processes. See <a href="https://www.nextflow.io/docs/latest/operator.html">here</a>
for details on the specific operators.</p>
</div>
<div class="paragraph">
<p>Before we perform the GATK ASEReadCounter process, we must group the data for allele-specific expression. To do this we must combine channels.</p>
</div>
<div class="paragraph">
<p>The <code>bam_for_ASE_ch</code> channel emites tuples having the following structure, holding the final BAM/BAI files:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-bash hljs" data-lang="bash">( sample_id, file_bam, file_bai )</code></pre>
</div>
</div>
<div class="paragraph">
<p>The <code>vcf_for_ASE</code> channel emits tuples having the following structure:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-bash hljs" data-lang="bash">( sample_id, output.vcf )</code></pre>
</div>
</div>
<div class="paragraph">
<p>In the first operation, the BAMs are grouped together by sample id.</p>
</div>
<div class="paragraph">
<p>Next, this resulting channel is merged with the VCFs (vcf_for_ASE) having the same sample id.</p>
</div>
<div class="paragraph">
<p>We must take the merged channel and creates a channel named <code>grouped_vcf_bam_bai_ch</code> emitting the following tuples:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-bash hljs" data-lang="bash">( sample_id, file_vcf, List[file_bam], List[file_bai] )</code></pre>
</div>
</div>
<div class="paragraph">
<p>Your aim is to fill in the <code>BLANKS</code> below.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bam_for_ASE_ch
  .BLANK                            <i class="conum" data-value="1"></i><b>(1)</b>
  .phase(vcf_for_ASE)               <i class="conum" data-value="2"></i><b>(2)</b>
  .map{ left, right -&gt;              <i class="conum" data-value="3"></i><b>(3)</b>
    def sampleId = left[0]          <i class="conum" data-value="4"></i><b>(4)</b>
    def bam = left[1]               <i class="conum" data-value="5"></i><b>(5)</b>
    def bai = left[2]               <i class="conum" data-value="6"></i><b>(6)</b>
    def vcf = right [1]             <i class="conum" data-value="7"></i><b>(7)</b>
    tuple(BLANK, vcf, BLANK, BLANK) <i class="conum" data-value="8"></i><b>(8)</b>
  .set { grouped_vcf_bam_bai_ch }   <i class="conum" data-value="9"></i><b>(9)</b></pre>
</div>
</div>
<div class="colist arabic">
<table>
<tr>
<td><i class="conum" data-value="1"></i><b>1</b></td>
<td>an operator that groups sets that contain a common first element.</td>
</tr>
<tr>
<td><i class="conum" data-value="2"></i><b>2</b></td>
<td>the phase operator synchronizes the values emitted by two other channels. See <a href="https://www.nextflow.io/docs/latest/operator.html?phase#phase">here</a> for more details</td>
</tr>
<tr>
<td><i class="conum" data-value="3"></i><b>3</b></td>
<td>the map operator can apply any function to every item on a channel. In this case we take our tuple from the phase operation, define the seperate elements and create a new tuple.</td>
</tr>
<tr>
<td><i class="conum" data-value="4"></i><b>4</b></td>
<td>define repID to be the first element of left.</td>
</tr>
<tr>
<td><i class="conum" data-value="5"></i><b>5</b></td>
<td>define bam to be the second element of left.</td>
</tr>
<tr>
<td><i class="conum" data-value="6"></i><b>6</b></td>
<td>define bai to be the third element of left.</td>
</tr>
<tr>
<td><i class="conum" data-value="7"></i><b>7</b></td>
<td>define vcf to be the first element of right.</td>
</tr>
<tr>
<td><i class="conum" data-value="8"></i><b>8</b></td>
<td>create a new tuple made of four elements</td>
</tr>
<tr>
<td><i class="conum" data-value="9"></i><b>9</b></td>
<td>rename the resulting as <code>grouped_vcf_bam_bai_ch</code></td>
</tr>
</table>
</div>
<div class="admonitionblock caution">
<table>
<tr>
<td class="icon">
<i class="fa icon-caution" title="Caution"></i>
</td>
<td class="content">
<code>left</code> and <code>right</code> above are arbitary names. From the phase operator documentation, we see that phase returns pairs of items. So here <code>left</code> originates from contents of the <code>bam_for_ASE_ch</code> channel and <code>right</code> originates from the contents of <code>vcf_for_ASE</code> channel.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><a href="solutions/kind_koala.html">Solution</a></p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_process_6c">4.12. Process 6C</h3>
<h3 id="_allele_specific_expression_analysis_with_gatk_asereadcounter" class="discrete">Allele-Specific Expression analysis with GATK ASEReadCounter</h3>
<div class="paragraph">
<p>Now we are ready for the final process.</p>
</div>
<div class="paragraph">
<p>You should implement a process having the following structure:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Name</dt>
<dd>
<p>6C_ASE_knownSNPs</p>
</dd>
<dt class="hdlist1">Command</dt>
<dd>
<p>calculate allele counts at a set of positions with GATK tools</p>
</dd>
<dt class="hdlist1">Input</dt>
<dd>
<p>genome fasta file<br>
genome index file from samtools<br>
genome dictionary file<br>
the `grouped_vcf_bam_bai_ch`channel</p>
</dd>
<dt class="hdlist1">Output</dt>
<dd>
<p>the allele specific expression file (<code>ASE.tsv</code>)</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_problem_12">4.12.1. Problem #12</h4>
<div class="paragraph">
<p>You should construct the process and run the pipeline in its entirety.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>  echo "${bam.join('\n')}" &gt; bam.list

  java -jar $GATK -R ${genome} \
                  -T ASEReadCounter \
                  -o ASE.tsv \
                  -I bam.list \
                  -sites ${vcf}</pre>
</div>
</div>
<div class="paragraph">
<p><a href="solutions/laughing_lynx.html">Solution</a></p>
</div>
<div class="paragraph">
<p>Congratulations! If you made it this far you now have all the basics to create your own Nextflow workflows.</p>
</div>
<hr>
</div>
</div>
<div class="sect2">
<h3 id="_results_overview">4.13. Results overview</h3>
<div class="paragraph">
<p>For each processed sample the pipeline stores results into a folder named after the sample identifier. These folders are created in the directory specified as a parameter in <code>params.results</code>.</p>
</div>
<div class="paragraph">
<p>Result files for this workshop can be found in the folder <code>results</code> within the current folder. There you should see a directory called <code>ENCSR000COQ/</code> containing the following files:</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Variant calls</dt>
<dd>
<p><code>final.vcf</code></p>
<div class="paragraph">
<p>This file contains all somatic variants (SNVs) called from RNAseq data. You will see variants that pass all filters, with the <code>PASS</code> keyword in the <span class="red">7th</span> field of the vcf file (<code>filter status</code>), and also those that did not pass one or more filters.</p>
</div>
<div class="paragraph">
<p><code>commonSNPs.diff.sites_in_files</code></p>
</div>
<div class="paragraph">
<p>Tab-separated file with comparison between variants obtained from RNAseq and "known" variants from DNA.</p>
</div>
<div class="paragraph">
<p>The file is sorted by genomic position and contains 8 fields:</p>
</div>
<table class="tableblock frame-all grid-all">
<colgroup>
<col>
<col>
<col>
</colgroup>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">1</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> CHROM</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">chromosome name;</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">2</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> POS1</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">position of the SNV in file #1 (RNAseq data);</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">3</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> POS2</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">position of SNV in file #2 (DNA "known" variants);</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">4</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> IN_FILE</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">flag whether SNV is present in the file #1 '1', in the file #2 '2', or in both files 'B';</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">5</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> REF1</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">reference sequence in the file 1;</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">6</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> REF2</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">reference sequence in the file 2;</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">7</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> ALT1</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">alternative sequence in the file 1;</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">8</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> ALT2</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">alternative sequence in the file 2</p></td>
</tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="paragraph">
<p><code>known_snps.vcf</code></p>
</div>
<div class="paragraph">
<p>Variants that are common to RNAseq and "known" variants from DNA.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1">Allele specific expression quantification</dt>
<dd>
<p><code>ASE.tsv</code></p>
<div class="paragraph">
<p>Tab-separated file with allele counts at common SNVs positions (only SNVs from the file <code>known_snps.vcf</code>)</p>
</div>
<div class="paragraph">
<p>The file is sorted by coordinates and contains 13 fields:</p>
</div>
<table class="tableblock frame-all grid-all">
<colgroup>
<col>
<col>
<col>
</colgroup>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">1</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> contig</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">contig, scaffold or chromosome name of the variant</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">2</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> position</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">position of the variant</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">3</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> variant ID</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">variant ID in the dbSNP</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">4</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> refAllele</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">reference allele sequence</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">5</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> altAllele</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">alternate allele sequence</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">6</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> refCount</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that support the reference allele</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">7</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> altCount</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that support the alternate allele</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">8</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> totalCount</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">total number of reads at the site that support both reference and alternate allele and any other alleles present at the site</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">9</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> lowMAPQDepth</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that have low mapping quality</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">10</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> lowBaseQDepth</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that have low base quality</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">11</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> rawDepth</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">total number of reads at the site that support both reference and alternate allele and any other alleles present at the site</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">12</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> otherBases</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that support bases other than reference and alternate bases</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock"><span class="red">13</span></p></td>
<td class="tableblock halign-center valign-top"><div class="literal"><pre> improperPairs</pre></div></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">number of reads that have malformed pairs</p></td>
</tr>
</tbody>
</table>
</dd>
<dt class="hdlist1">Allele frequency histogram</dt>
<dd>
<p><code>AF.histogram.pdf</code></p>
<div class="paragraph">
<p>This file contains a histogram plot of allele frequency for SNVs common to RNA-seq and "known" variants from DNA.</p>
</div>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_bonus_step">4.14. Bonus step</h3>
<div class="paragraph">
<p>Until now the pipeline has been executed using just a single sample (<code>ENCSR000COQ1</code>).</p>
</div>
<div class="paragraph">
<p>Now we can re-execute the pipeline specifying a large set of samples by using the command
shown below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlightjs highlight"><code class="language-cmd hljs" data-lang="cmd">nextflow run main.nf -resume --reads 'data/reads/ENCSR000C*_{1,2}.fastq.gz'</code></pre>
</div>
</div>
<div class="paragraph">
<p>It will print an output similar to the one below:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>N E X T F L O W  ~  version 0.24.1
Launching `main.nf` [backstabbing_nightingale] - revision: 1187e44c7a
[warm up] executor &gt; local
[c6/75e3f4] Submitted process &gt; 1A_prepare_genome_samtools (genome)
[7b/44e5d6] Submitted process &gt; 1C_prepare_star_genome_index (genome)
[da/e19bcf] Submitted process &gt; 1B_prepare_genome_picard (genome)
[95/1ad13d] Submitted process &gt; 1D_prepare_vcf_file (known_variants.vcf)
[72/702900] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000COR1)
[9a/5ca042] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000CPO1)
[77/03ef01] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000COR2)
[04/262db9] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000COQ2)
[a4/64c69e] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000CPO2)
[9e/ad3621] Submitted process &gt; 2_rnaseq_mapping_star (ENCSR000COQ1)
[a5/cda1b0] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000COQ2)
[42/0565d7] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000COQ1)
[0c/68ce48] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000COR1)
[6b/3843e1] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000COR2)
[1c/8c474b] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000CPO1)
[98/f17992] Submitted process &gt; 3_rnaseq_gatk_splitNcigar (ENCSR000CPO2)
[c2/8cdfca] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000COQ1)
[d1/1a6935] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000COR1)
[9f/b4c61d] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000COR2)
[aa/b43a43] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000COQ2)
[46/2d96f0] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000CPO1)
[85/6b9527] Submitted process &gt; 4_rnaseq_gatk_recalibrate (ENCSR000CPO2)
[79/a7fb48] Submitted process &gt; 5_rnaseq_call_variants (ENCSR000CPO)
[a5/29c017] Submitted process &gt; 5_rnaseq_call_variants (ENCSR000COQ)
[22/1fdea2] Submitted process &gt; 5_rnaseq_call_variants (ENCSR000COR)
[7d/e1adfb] Submitted process &gt; 6A_post_process_vcf (ENCSR000CPO)
[0a/4d43fc] Submitted process &gt; 6A_post_process_vcf (ENCSR000COQ)
[18/8d486b] Submitted process &gt; 6A_post_process_vcf (ENCSR000COR)
[60/427153] Submitted process &gt; 6B_prepare_vcf_for_ase (ENCSR000CPO)
[32/64eff0] Submitted process &gt; 6B_prepare_vcf_for_ase (ENCSR000COQ)
[31/32ad40] Submitted process &gt; 6B_prepare_vcf_for_ase (ENCSR000COR)
[6f/a5e211] Submitted process &gt; 6C_ASE_knownSNPs (ENCSR000COR)
[ff/989dc1] Submitted process &gt; 6C_ASE_knownSNPs (ENCSR000CPO)
[25/92875a] Submitted process &gt; 6C_ASE_knownSNPs (ENCSR000COQ)</pre>
</div>
</div>
<div class="paragraph">
<p>You can notice that this time the pipeline spawns the execution of more tasks because
three samples have been provided instead of one.</p>
</div>
<div class="paragraph">
<p>This shows the ability of Nextflow to implicitly handle multiple parallel task executions
depending on the specified pipeline input dataset.</p>
</div>
</div>
</div>
</div>
</div>
<script type="text/javascript">
    var tocitems = $('#toc li a'),
    scrollItems = tocitems.map(function(){
      var item = $($(this).attr("href"));
      if (item.length) { return item; }
    });
    $(document).ready(function() {
        var lastId
        $(window).scroll(function() {
              // Get container scroll position
            var fromTop = $(this).scrollTop();

            // Get id of current scroll item
            var cur = scrollItems.map(function(){
                var offset = this.offset().top - parseFloat(this.css('margin-top'))
                if (offset < fromTop) {
                    return this;
                }
            });
            // Get the id of the current element
            cur = cur[cur.length-1];
            var id = cur && cur.length ? cur[0].id : "";

            if (lastId !== id) {
               lastId = id;
               // Set/remove active class
               scrollItems
                 .end().parent()
                 .removeClass("active")
                 .end().filter("[href='#"+id+"']")
                 .parent().addClass("active");
            }
        } );
    } );
</script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlighting()</script>
</body>
</html>